### PR TITLE
User Interactions data collection and download

### PR DESCRIPTION
**Pull request by imsickofmaps**
## Basics

This is the only POST API endpoint at the moment. This is to allow for data collection and then download by the admin. 
- This app should be called "User Interactions".
- It will have one POST endpoint for inserting data into the database
- The admin should be able to browse the data using the usual listing views in Django
- The admin should be able to click a button and download all the data in CSV format
### POST API

The post needs to accept the following data:
- msisdn (Max 20 chars)
- action (Max 200 characters) <- REMOVE
- feature (Max 30 chars) <-- ADD
- key (Max 100 chars) <-- ADD
- value (Max 200 chars) <-- ADD
- transport (Max 5 characters)
- created_at (Auto-generated on insert)

All fields are required.

### CODE REVIEW

**Comment by imsickofmaps** [comment]
@gitumarkk - see the details above - can you change the model and the resulting API please

File: user_interactions/admin.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from django.contrib import admin
+from models import UserInteraction
+from actions import export_as_csv_action
+
+
+class UserInteractionAdmin(admin.ModelAdmin):
+    list_display = ["msisdn", "transport", "created_at"]
+    actions = [export_as_csv_action("CSV Export",
+                                    fields=["msisdn", "action",
```
Comment (imsickofmaps) [inline user_interactions/admin.py]: Missing new fields

File: user_interactions/views.py [hunk 2]
```diff
@@ -0,0 +1 @@
+# Create your views here.
```
Comment (imsickofmaps) [inline user_interactions/views.py]: just delete this please

Comment (imsickofmaps) [inline user_interactions/views.py]: the file that is

File: user_interactions/api.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    def post_list(self, request, **kwargs):
+        """
+        Gets posted data so as to save it in the UserInteraction Table, but
+        checks if the characters are not too long (if this is not done TastyPie
+        returns a huge annoying error)
+        """
+        user_data = request.POST.dict()
+        if (len(user_data['msisdn']) > 20):
+            raise TastypieError("msisdn Chars are too long, len = %s"
```
Comment (imsickofmaps) [inline user_interactions/api.py]: can we just make this and the other errors below "'fieldname' is longer than the maximim allowed length of 'length'"

File: user_interactions/actions.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import csv
+from django.http import HttpResponse
+
+
+def export_as_csv_action(description="Export selected objects as CSV file",
```
Comment (imsickofmaps) [inline user_interactions/actions.py]: Please change description to "Export selected records to CSV file"

**Comment by imsickofmaps** [comment]
@smn ready for review

File: user_interactions/tests/test_api.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+
+class TestUserInteraction(TestCase):
+
+    def test_api_works(self):
+        url = reverse('api_dispatch_list',
+                      kwargs={'resource_name': 'userinteraction',
+                      'api_name': 'api'})
+        response = self.client.post(url,
+                                    format="json",
```
Comment (smn) [inline user_interactions/tests/test_api.py:16]: If you specify `Content-Type: application/json` in your request header you can omit this.

File: user_interactions/tests/test_api.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+                                    data={"msisdn": "post msisdn",
+                                    "feature": "post feature",
+                                    "key": "post key",
+                                    "value": "post value",
+                                    "transport": "sms"})
```
Comment (smn) [inline user_interactions/tests/test_api.py:21]: I'm not sure this is automatically serialized as JSON for you.

File: user_interactions/tests/test_api.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        query = UserInteraction.objects.all()
+        self.assertEqual(response.status_code, 204)
+        self.assertEqual(query[0].msisdn, "post msisdn")
```
Comment (smn) [inline user_interactions/tests/test_api.py:24]: Are you sure there's only 1 `UserInteraction` record?
Going with the following is a bit safer:

``` python
ui = UserInteraction.objects.latest('pk')
self.assertEqual(ui.msisdn, "post msisdn")
```

File: user_interactions/tests/test_api.py (hunk 6 above)
Comment (smn) [inline user_interactions/tests/test_api.py:21]: So it looks like we're just doing a regular HTTP POST and expecting JSON back as a response. Ignore my previous 2 comments on this.

**Comment by smn** [comment]
:+1:

**Pull request merged**